### PR TITLE
Install improvement

### PR DESCRIPTION
**Pull request by jasmingacic**
This improvement allows to pass all possible k0s parameters to install command. 
Syntax is like this:

```
# ./k0s install --role server --enable-worker
INFO[2020-12-31 13:12:53] Installing k0s service                       
# cat /etc/systemd/system/k0s.service 
[Unit]
Description=k0s - Zero Friction Kubernetes
ConditionFileIsExecutable=/root/go/src/github.com/k0sproject/k0s/k0s

After=network.target 

[Service]
StartLimitInterval=5
StartLimitBurst=10
ExecStart=/root/go/src/github.com/k0sproject/k0s/k0s "server" "--config" "/root/go/src/github.com/k0sproject/k0s/k0s.yaml" "--enable-worker"
...
```

I believe this will clear up the questions related to #593  or similar

p.s. as @jnummelin mentioned maybe it would be worth considering adding separate install commands for server and worker so it would be easier to understand.

### CODE REVIEW

File: cmd/install.go [hunk 1]
```diff
@@ -108,19 +108,57 @@ func getUserList(sysUsers v1beta1.SystemUser) []string {
...
+		//server params
+		if serverWorkerProfile != "default" {
+			params["profile"] = serverWorkerProfile
+		}
+
+		if enableWorker {
+			params["enable-worker"] = ""
+		}
+
+		params["token-file"] = serverToken
```
Comment (kke) [inline cmd/install.go:129]: I don't see where this `serverToken` comes from

File: cmd/install.go [hunk 2]
```diff
@@ -108,19 +108,57 @@ func getUserList(sysUsers v1beta1.SystemUser) []string {
...
+
 		if cfgFile != "" {
 			args = append(args, "--config", cfgFile)
 		}
 	} else {
-		args = append(args, "worker", "--token-file", "REPLACEME")
+		args = append(args, "worker")
+
+		//worker params
+		if workerProfile != "default" {
```
Comment (kke) [inline cmd/install.go:138]: Same with this, where does `workerProfile` come from?

File: cmd/install.go [hunk 3]
```diff
@@ -108,19 +108,57 @@ func getUserList(sysUsers v1beta1.SystemUser) []string {
...
+			params["profile"] = workerProfile
```
Comment (kke) [inline cmd/install.go:139]: Same with this, where does `workerProfile` come from?

File: cmd/install.go (hunk 1 above)
Comment (jasmingacic) [reply cmd/install.go:129]: it comes from https://github.com/k0sproject/k0s/blob/main/cmd/server.go#L59

File: cmd/install.go [hunk 4]
```diff
@@ -108,19 +108,57 @@ func getUserList(sysUsers v1beta1.SystemUser) []string {
...
+		}
+
+		if tokenFile != "" {
+			params["token-file"] = tokenFile
+		} else {
+			params["token-file"] = "REPLACEME"
+
+		}
+		params["cri-socket"] = criSocket
```
Comment (kke) [inline cmd/install.go:148]: Same with these, where are `workerProfile`, `tokenFile` and `criSocket` coming from?

File: cmd/install.go (hunk 1 above)
Comment (kke) [reply cmd/install.go:129]: How does it appear in `cmd/install.go`'s `getCmdArgs` function scope?

File: cmd/install.go (hunk 4 above)
Comment (jasmingacic) [reply cmd/install.go:148]: https://github.com/k0sproject/k0s/blob/main/cmd/worker.go#L42

File: cmd/install.go (hunk 1 above)
Comment (kke) [reply cmd/install.go:129]: image

**Comment by jasmingacic** [comment]
The most recent changes are:
```
# ./k0s install -h
Helper command for setting up k0s on a brand-new system. Must be run as root (or with sudo)
Usage:
  k0s install [command]
Available Commands:
  server      Helper command for setting up k0s as server node on a brand-new system. Must be run as root (or with sudo)
  worker      Helper command for setting up k0s as a worker node on a brand-new system. Must be run as root (or with sudo)
```
Server subcommand:
```
# ./k0s install server -h
Helper command for setting up k0s as server node on a brand-new system. Must be run as root (or with sudo)

Usage:
  k0s install server [flags]

Examples:
All default values of server command will be passed to the service stub unless overriden. 
To get information about default values run "k0s install server --help"

With server subcommand you can setup a single node cluster by running:

        k0s install server --enable-worker

```
Worker subcommand:
```
# ./k0s install worker -h
Helper command for setting up k0s as a worker node on a brand-new system. Must be run as root (or with sudo)

Usage:
  k0s install worker [flags]

Examples:
Worker subcommand allows you to pass in all available worker parameters. 
All default values of worker command will be passed to the service stub unless overriden.
To get information about default values run "k0s install server --help"

Note: if --token-file value is not passed service stub will add "REPLACEME" value instead of it.
Windows flags like "--api-server", "--cidr-range" and "--cluster-dns" will be ignores since install command doesn't yet support Windows services

Flags:
      --api-server string       HACK: api-server for the windows worker node
      --cidr-range string       HACK: cidr range for the windows worker node (default "10.96.0.0/12")
      --cluster-dns string      HACK: cluster dns for the windows worker node (default "10.96.0.10")
      --cri-socket string       contrainer runtime socket to use, default to internal containerd. Format: [remote|docker]:[path-to-socket]
      --enable-cloud-provider   Whether or not to enable cloud provider support in kubelet
  -h, --help                    help for worker
      --profile string          worker profile to use on the node (default "default")
      --token-file string       Path to the file containing token.

```
Install server with default values:
```
# ./k0s install server
INFO[2021-01-04 13:50:14] Installing k0s service                       
# cat /etc/systemd/system/k0sserver.service 
[Unit]
Description=k0s - Zero Friction Kubernetes
ConditionFileIsExecutable=/root/go/src/github.com/k0sproject/k0s/k0s
After=network.target 
[Service]
StartLimitInterval=5
StartLimitBurst=10
ExecStart=/root/go/src/github.com/k0sproject/k0s/k0s "server" "--config" "/root/go/src/github.com/k0sproject/k0s/k0s.yaml"
```
Install worker with default values:
```
# ./k0s install worker --token-file whatever
INFO[2021-01-04 13:51:25] Installing k0s service                       
root@k0s:~/go/src/github.com/k0sproject/k0s# cat /etc/systemd/system/k0sworker.service 
[Unit]
Description=k0s - Zero Friction Kubernetes
ConditionFileIsExecutable=/root/go/src/github.com/k0sproject/k0s/k0s
After=network.target 
[Service]
StartLimitInterval=5
StartLimitBurst=10
ExecStart=/root/go/src/github.com/k0sproject/k0s/k0s "worker" "--token-file whatever"
```

**Comment by jasmingacic** [comment]
One major change is that service name will be set according to k0s role. So when installing k0s server the service file will be names k0sserver.service and k0sworker.service respectively.

File: pkg/install/service.go [hunk 5]
```diff
@@ -30,14 +30,16 @@ func EnsureService(args []string) error {
 	var k0sDisplayName string
 
 	prg := &program{}
-
 	for _, v := range args {
 		if v == "server" {
 			k0sDisplayName = "k0s server"
+			k0sServiceName = k0sServiceName + v
+		} else if v == "worker" {
```
Comment (kke) [inline pkg/install/service.go:37]: Can this somehow go to the `else` branch? 

worker` doesn't seem to get a `k0sDisplayName` like before.

Looks to me like the whole conditional could be skipped and replaced with:
```go
for _, v := range args {
  k0sDisplayName = "k0s " + v
  k0sServiceName = "k0s" + v
}
```

I have no clue what it is doing, but looks weird.

Comment (kke) [reply pkg/install/service.go:37]: Or maybe inside `if v == "server" || v == "worker"`

Comment (jasmingacic) [reply pkg/install/service.go:37]: Display name and service name are not the same. 
Service name is the name of  the service stub like k0sserver.service or k0sworker.service where the display name is just how it is going to appear when listed with systemctl.
p.s. Args slice contains more things than just role name so it had to have the condition as above

File: go.mod [hunk 6]
```diff
@@ -36,6 +35,8 @@ require (
 	github.com/spf13/cobra v1.1.1
 	github.com/spf13/viper v1.7.1
 	github.com/stretchr/testify v1.6.1
+	github.com/vishvananda/netlink v1.1.0 // indirect
+	github.com/vishvananda/netns v0.0.0-20201230012202-c4f3ca719c73 // indirect
```
Comment (kke) [inline go.mod:39]: Seem to be unrelated

File: go.sum [hunk 7]
```diff
@@ -800,6 +794,11 @@ github.com/urfave/cli v1.22.1/go.mod h1:Gos4lmkARVdJ6EkW0WaNv/tZAAMe9V7XWyB60NtX
 github.com/valyala/bytebufferpool v1.0.0/go.mod h1:6bBcMArwyJ5K/AmCkWv1jt77kVWyCJ6HpOuEn7z0Csc=
 github.com/valyala/fasttemplate v1.0.1/go.mod h1:UQGH1tvbgY+Nz5t2n7tXsz52dQxojPUpymEIMZ47gx8=
 github.com/vektah/gqlparser v1.1.2/go.mod h1:1ycwN7Ij5njmMkPPAOaRFY4rET2Enx7IkVv3vaXspKw=
+github.com/vishvananda/netlink v1.1.0 h1:1iyaYNBLmP6L0220aDnYQpo1QEV4t4hJ+xEEhhJH8j0=
+github.com/vishvananda/netlink v1.1.0/go.mod h1:cTgwzPIzzgDAYoQrMm0EdrjRUBkTqKYppBueQtXaqoE=
+github.com/vishvananda/netns v0.0.0-20191106174202-0a2b9b5464df/go.mod h1:JP3t17pCcGlemwknint6hfoeCVQrEMVwxRLRjXpq+BU=
+github.com/vishvananda/netns v0.0.0-20201230012202-c4f3ca719c73 h1:JGMFiSX7pNu3l0DRO8sjxgA7ybQNE+HeuKIG23PjsN4=
+github.com/vishvananda/netns v0.0.0-20201230012202-c4f3ca719c73/go.mod h1:DD4vA1DwXk04H54A1oHXtwZmA0grkVMdPxx/VGLCah0=
```
Comment (kke) [inline go.sum:801]: Seem to be unrelated

Comment (jasmingacic) [reply go.sum:801]: I did `go mod tidy` so I haven't touched it

File: go.mod (hunk 6 above)
Comment (jasmingacic) [reply go.mod:39]: I did `go mod tidy` so I haven't touched it

File: pkg/install/service.go (hunk 5 above)
Comment (kke) [reply pkg/install/service.go:37]: Probably doesn't matter, but maybe it should break after the first match.

File: go.mod (hunk 6 above)
Comment (jasmingacic) [reply go.mod]: Weirdly `go mod tidy` actually adds these here

File: cmd/install.go [hunk 8]
```diff
@@ -108,19 +94,58 @@ func getUserList(sysUsers v1beta1.SystemUser) []string {
...
 
 	if debug {
 		args = append(args, "--debug")
 	}
+
+	for k, v := range params {
+		if v != "" {
+			args = append(args, (fmt.Sprintf("--%s %s", k, v)))
+		}
+		if k == "enable-worker" || k == "enable-cloud-provider" {
```
Comment (kke) [inline cmd/install.go:145]: Couldn't this just be:

```suggestion
		} else {
```

Comment (jasmingacic) [reply cmd/install.go:145]: yes it could

File: pkg/install/service.go (hunk 5 above)
Comment (kke) [reply pkg/install/service.go:38]: Could have been simply
```go
  for _, v := range args {
    if v == "server" || v == "worker" {
      k0sDisplayName = "k0s " + v
      k0sServiceName = k0sServiceName + v
      break
    }
  }
```

Comment (jasmingacic) [reply pkg/install/service.go]: Fixed it

File: cmd/install.go (hunk 8 above)
Comment (kke) [inline cmd/install.go:145]: Appears `params['enable-cloud-provider']` will never be set.

**Review by kke — CHANGES_REQUESTED** [review]

File: cmd/install.go (hunk 8 above)
Comment (jasmingacic) [reply cmd/install.go:145]: It has default value set to false 
`workerCmd.Flags().BoolVar(&cloudProvider, "enable-cloud-provider", false, "Whether or not to enable cloud provider support in kubelet")`

Such flags like enable-worker or enable-cloud-provider are by default set to false but when they are passed like `k0s worker --enable-cloud-provider` then they are set to true.

Comment (jasmingacic) [reply cmd/install.go]: Which I never set in the block before. Fixed it now.

File: cmd/install.go [hunk 9]
```diff
@@ -108,19 +94,61 @@ func getUserList(sysUsers v1beta1.SystemUser) []string {
...
+		params["cri-socket"] = criSocket
 	}
 
 	if debug {
 		args = append(args, "--debug")
 	}
+
+	for k, v := range params {
+		if v != "" {
+			args = append(args, (fmt.Sprintf("--%s %s", k, v)))
```
Comment (kke) [inline cmd/install.go:145]: Extra parenthesis around `fmt.Sprintf`

File: cmd/install.go [hunk 10]
```diff
@@ -108,19 +94,61 @@ func getUserList(sysUsers v1beta1.SystemUser) []string {
...
+		}
+
+		if k == "enable-worker" || k == "enable-cloud-provider" {
+			args = append(args, (fmt.Sprintf("--%s", k)))
+
+		}
+	}
```
Comment (kke) [inline cmd/install.go:152]:
```suggestion
	for k, v := range params {
		arg := "--" + k
		if v != "" {
			arg += "=" + v
		}
		args = append(args, arg)
	}
```

I don't see a reason to identify the `enable-worker` and `enable-cloud-provider` keys here. It will require an extra change you have to make if you add any other flag without a value.

The intermediate `map[string]string` is a bit unnecessary too, the whole function could just simply be:

```go
func getCmdArgs(role string) args []string {
	args = append(args, role)
	args = append(args, "--data-dir", dataDir)
	if role == "server" {
		args = append(args, "--token-file", serverToken)

		if serverWorkerProfile != "default" {
			args = append(args, "--profile" , serverWorkerProfile)
		}

		if enableWorker {
			args = append(args, "--enable-worker")
		}

		if cfgFile != "" {
			args = append(args, "--config", cfgFile)
		}
	} else {
		args = append(args, "--cri-socket", criSocket)

		if workerProfile != "default" {
			args = append(args, "--profile", workerProfile)
		}

		if tokenFile != "" {
			args = append(args, "--token-file", tokenFile)
		} else {
			args = append(args, "--token-file", "REPLACEME")
		}

		if cloudProvider {
			args = append(args, "--enable-cloud-provider")
		}
	}

	if debug {
		args = append(args, "--debug")
	}

	return args
}
```

Shorter, simpler, less intermediate allocation, only uses anonymous string literals (=constants)

File: cmd/installWorker.go [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+	"github.com/spf13/cobra"
+)
+
+var (
+	installWorkerCmd = &cobra.Command{
+		Use:   "worker",
+		Short: "Helper command for setting up k0s as a worker node on a brand-new system. Must be run as root (or with sudo)",
+		Example: `Worker subcommand allows you to pass in all available worker parameters. 
+All default values of worker command will be passed to the service stub unless overriden.
+To get information about default values run "k0s install server --help"
```
Comment (kke) [inline cmd/installWorker.go:28]:
```suggestion
To get information about default values run "k0s install worker --help"
```

File: cmd/installWorker.go [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+
+Note: if --token-file value is not passed service stub will add "REPLACEME" value instead of it.
```
Comment (kke) [inline cmd/installWorker.go:30]:
```suggestion
Note: if --token-file value is not passed, service stub will add "REPLACEME" value instead of it.
```

**Review by kke — DISMISSED** [review]
I don't see why the temporary `params map[string]string` construct in `cmd/install.go` is needed, I wrote a full replacement for this function without that in a [review comment](https://github.com/k0sproject/k0s/pull/594#discussion_r553805473):

```go
func getCmdArgs(role string) args []string {
	args = append(args, role)
	args = append(args, "--data-dir", dataDir)
	if role == "server" {
		args = append(args, "--token-file", serverToken)

		if serverWorkerProfile != "default" {
			args = append(args, "--profile" , serverWorkerProfile)
		}

		if enableWorker {
			args = append(args, "--enable-worker")
		}

		if cfgFile != "" {
			args = append(args, "--config", cfgFile)
		}
	} else if role == "worker" {
		args = append(args, "--cri-socket", criSocket)

		if workerProfile != "default" {
			args = append(args, "--profile", workerProfile)
		}

		if tokenFile != "" {
			args = append(args, "--token-file", tokenFile)
		} else {
			args = append(args, "--token-file", "REPLACEME")
		}

		if cloudProvider {
			args = append(args, "--enable-cloud-provider")
		}
	}

	if debug {
		args = append(args, "--debug")
	}

	return
}
```

Other than that, LGTM.

File: cmd/install.go (hunk 10 above)
Comment (jasmingacic) [reply cmd/install.go]: @kke can you please re-review this file. I added requested changes.

**Review by kke — DISMISSED** [review]

File: cmd/install.go [hunk 13]
```diff
@@ -107,20 +93,50 @@ func getUserList(sysUsers v1beta1.SystemUser) []string {
...
+	args = append(args, role)
+
+	if dataDir != "" {
+		args = append(args, "--data-dir", dataDir)
+	}
 
 	if role == "server" {
-		args = append(args, "server")
+		if serverToken != "" {
+			args = append(args, "--token-file", serverToken)
```
Comment (trawler) [inline cmd/install.go:105]: shouldn't `serverToken` actually be changed to `tokenFile`?

**Review by trawler — CHANGES_REQUESTED** [review]
- please also update https://github.com/k0sproject/k0s/blob/main/docs/install.md
- You have some unrelated commits bundled in this PR. Please rebase and clean up your branch before merging. It makes it a bit hard to review and pinpoint the relevant changes.

File: cmd/install.go (hunk 13 above)
Comment (jasmingacic) [reply cmd/install.go:105]: The variable is inherited from here https://github.com/k0sproject/k0s/blob/main/cmd/server.go#L59 so it is already parsed.

Comment (trawler) [reply cmd/install.go:105]: I believe this is the argument you are looking for: https://github.com/k0sproject/k0s/blob/main/cmd/server.go#L53

Comment (jasmingacic) [reply cmd/install.go]: Fixed it, can you please re-review it.

**Comment by jasmingacic** [comment]
I wonder if we should change status of this command from alpha to beta? https://github.com/k0sproject/k0s/blob/main/docs/install.md

**Comment by jasmingacic** [comment]
Rebased and addressed all the remarks

**Review by jnummelin — APPROVED** [review]
LGTM

**Review by soider — APPROVED** [review]
good stuff
